### PR TITLE
Kfspts 3599-new-4

### PR DESCRIPTION
**Pull request by nkimble**
Latest round of requested code review changes.

### CODE REVIEW

**Comment by nkimble** [comment]
retest this please

**Comment by db483** [comment]
Changes look good. If Bryan signs off then I can merge it once you squash the commits into one. Thank you.

File: src/main/java/edu/cornell/kfs/coa/businessobject/SubAccountGlobal.java [hunk 1]
```diff
@@ -121,83 +126,63 @@ public SubAccountGlobal() {
...
+		    updateIcrAccounts(subAccountGlobalDetail, icrAccounts);
 		    
 			changesToPersist.add(subAccount);
 		}
 		
 		return changesToPersist;
 	}
+	
+    public List<IndirectCostRecoveryAccountChange> getActiveIndirectCostRecoveryAccounts() {
+    	return SpringContext.getBean(GlobalObjectWithIndirectCostRecoveryAccountsService.class).getActiveIndirectCostRecoveryAccounts(this);
```
Comment (bhutchinson) [inline src/main/java/edu/cornell/kfs/coa/businessobject/SubAccountGlobal.java]: One more SpringContext.getBean() call that could be replaced by a getter/setter.

File: src/main/java/edu/cornell/kfs/coa/businessobject/SubAccountGlobal.java [hunk 2]
```diff
@@ -210,8 +195,24 @@ private A21IndirectCostRecoveryAccount createA21IndirectCostRecoveryAccountFromC
...
+		SubAccountGlobalDetail subAccountGlobalDetail = (SubAccountGlobalDetail)globalDetail;
+		List<A21IndirectCostRecoveryAccount> updatedA21IcrAccounts = new ArrayList<A21IndirectCostRecoveryAccount>();
+		for(IndirectCostRecoveryAccount icrAccount : updatedIcrAccounts){
+			updatedA21IcrAccounts.add(A21IndirectCostRecoveryAccount.copyICRAccount(icrAccount));
+		}
+		subAccountGlobalDetail.getSubAccount().getA21SubAccount().setA21IndirectCostRecoveryAccounts(updatedA21IcrAccounts);
+	}
+	
+	public void updateIcrAccounts(GlobalBusinessObjectDetailBase globalDetail, List<IndirectCostRecoveryAccount> icrAccounts){
+		SpringContext.getBean(GlobalObjectWithIndirectCostRecoveryAccountsService.class).updateIcrAccounts(this, globalDetail, icrAccounts);
```
Comment (bhutchinson) [inline src/main/java/edu/cornell/kfs/coa/businessobject/SubAccountGlobal.java]: Another more SpringContext.getBean() call that could be replaced by a getter/setter.

File: src/main/java/edu/cornell/kfs/coa/document/validation/impl/AccountGlobalPreRules.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        }
+
+    }
+    
+    @SuppressWarnings("deprecation")
+    protected boolean checkOffCampus() {
+      boolean continueRules = true;
+
+      if (accountGlobal.getAccountOffCampusIndicator() !=null && accountGlobal.getAccountOffCampusIndicator()) {
+        String questionText = SpringContext.getBean(ConfigurationService.class).getPropertyValueAsString(CUKFSKeyConstants.QUESTION_ACCOUNT_OFF_CAMPUS_INDICATOR);
```
Comment (bhutchinson) [inline src/main/java/edu/cornell/kfs/coa/document/validation/impl/AccountGlobalPreRules.java]: Another more SpringContext.getBean() call that could be replaced by a getter/setter.

**Comment by bhutchinson** [comment]
Changes generally look good, I just commented on three places where SpringContext.getBean() could be replaced with a getter/setter like you've already done elsewhere.

File: src/main/java/edu/cornell/kfs/coa/businessobject/SubAccountGlobal.java (hunk 1 above)
Comment (nkimble) [inline src/main/java/edu/cornell/kfs/coa/businessobject/SubAccountGlobal.java]: changed

File: src/main/java/edu/cornell/kfs/coa/businessobject/SubAccountGlobal.java (hunk 2 above)
Comment (nkimble) [inline src/main/java/edu/cornell/kfs/coa/businessobject/SubAccountGlobal.java]: changed

File: src/main/java/edu/cornell/kfs/coa/document/validation/impl/AccountGlobalPreRules.java (hunk 3 above)
Comment (nkimble) [inline src/main/java/edu/cornell/kfs/coa/document/validation/impl/AccountGlobalPreRules.java]: changed.

**Comment by nkimble** [comment]
If this round of changes are acceptable and the unit tests pass, I will rebase my changes onto Daniela's so there is a single commit for this PR.  Let me know.

File: src/main/java/edu/cornell/kfs/coa/businessobject/CuAccountGlobal.java [hunk 4]
```diff
@@ -2,164 +2,903 @@
...
     private static final long serialVersionUID = 1L;
 
-    protected String majorReportingCategoryCode;
-    protected MajorReportingCategory majorReportingCategory;
-    
-    @Override
-    public List<PersistableBusinessObject> generateGlobalChangesToPersist() {
+    private DateTimeService dateTimeService;
+    private BusinessObjectService businessObjectService;
+    private GlobalObjectWithIndirectCostRecoveryAccountsService globalObjectWithIndirectCostRecoveryAccountsService;
```
Comment (bhutchinson) [inline src/main/java/edu/cornell/kfs/coa/businessobject/CuAccountGlobal.java]: I would make these three services protected. I've seen issues in the past where the class variables themselves needed to be protected, having public getters/setters wasn't good enough.

File: src/main/java/edu/cornell/kfs/coa/businessobject/SubAccountGlobal.java [hunk 5]
```diff
@@ -14,29 +14,36 @@
...
 import org.kuali.rice.krad.bo.PersistableBusinessObjectBase;
 import org.kuali.rice.krad.service.PersistenceStructureService;
+import org.kuali.rice.krad.util.ObjectUtils;
+
+import edu.cornell.kfs.coa.service.GlobalObjectWithIndirectCostRecoveryAccountsService;
+import edu.cornell.kfs.sys.CUKFSPropertyConstants;
+
+
+public class SubAccountGlobal extends PersistableBusinessObjectBase implements GlobalObjectWithIndirectCostRecoveryAccounts {
+    private GlobalObjectWithIndirectCostRecoveryAccountsService globalObjectWithIndirectCostRecoveryAccountsService;
```
Comment (bhutchinson) [inline src/main/java/edu/cornell/kfs/coa/businessobject/SubAccountGlobal.java]: I would make this protected.

File: src/main/java/edu/cornell/kfs/coa/document/validation/impl/AccountGlobalPreRules.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+import org.kuali.rice.kns.document.MaintenanceDocument;
+import org.kuali.rice.krad.util.ObjectUtils;
+
+import edu.cornell.kfs.coa.businessobject.CuAccountGlobal;
+import edu.cornell.kfs.coa.businessobject.IndirectCostRecoveryAccountChange;
+import edu.cornell.kfs.sys.CUKFSKeyConstants;
+
+public class AccountGlobalPreRules extends MaintenancePreRulesBase {
+	
+    private ConfigurationService configurationService;
```
Comment (bhutchinson) [inline src/main/java/edu/cornell/kfs/coa/document/validation/impl/AccountGlobalPreRules.java]: I would make this protected.

**Comment by bhutchinson** [comment]
Just one more minor thing related to access modifiers. Thanks Nancy.

File: src/main/java/edu/cornell/kfs/coa/businessobject/CuAccountGlobal.java (hunk 4 above)
Comment (nkimble) [inline src/main/java/edu/cornell/kfs/coa/businessobject/CuAccountGlobal.java]: changed

File: src/main/java/edu/cornell/kfs/coa/businessobject/SubAccountGlobal.java (hunk 5 above)
Comment (nkimble) [inline src/main/java/edu/cornell/kfs/coa/businessobject/SubAccountGlobal.java]: changed

File: src/main/java/edu/cornell/kfs/coa/document/validation/impl/AccountGlobalPreRules.java (hunk 6 above)
Comment (nkimble) [inline src/main/java/edu/cornell/kfs/coa/document/validation/impl/AccountGlobalPreRules.java]: changed

**Comment by nkimble** [comment]
Same message as before. If this round of changes are acceptable and the unit tests pass, I will rebase my changes onto Daniela's so there is a single commit for this PR. Let me know.

**Comment by nkimble** [comment]
retest this please

**Comment by bhutchinson** [comment]
Looks good to me, thanks Nancy.

**Comment by nkimble** [comment]
I have rebased my changes onto Daniela's.

**Comment by nkimble** [comment]
retest this please

**Pull request merged**